docs/content/doc/developers/guidelines-refactoring.en-us.md

```diff
@@ -0,0 +1,52 @@
+---
+date: "2023-02-14T00:00:00+00:00"
+title: "Guidelines for Refactoring"
+slug: "guidelines-refactoring"
+weight: 20
+toc: false
+draft: false
+menu:
+sidebar:
+parent: "developers"
+name: "Guidelines for Refactoring"
+weight: 20
+identifier: "guidelines-refactoring"
+---
+
+# Guidelines for Refactoring
+
+**Table of Contents**
+
+{{< toc >}}
+
+## Background
+
+Since the first line of code was written at Feb 12, 2014, Gitea has grown to be a large project.
+As a result, the codebase has become larger and larger. The larger the codebase is, the more difficult it is to maintain.
+A lot of outdated mechanisms exist, a lot of frameworks are mixed together, some legacy code might cause bugs and block new features.
+To make the codebase more maintainable and make Gitea better, developers should keep in mind to use modern mechanisms to refactor the old code.
+
+This document is a collection of guidelines for refactoring the codebase.
+
+
+## Refactoring Suggestions
+
+* Design more about the future, but not only resolve the current problem.
+* Reduce ambiguity, reduce conflicts, improve maintainability.
+* Describe the refactoring, for example:
+  * Why the refactoring is needed.
+  * How the legacy problems would be solved.
+  * What's the Pros/Cons of the refactoring.
+* Only do necessary changes, keep the old logic as much as possible.
+* Introduce some intermediate steps to make the refactoring easier to review, a complete refactoring plan could be done in several PRs.
+* If there is any divergence, the TOC(Technical Oversight Committee) should be involved to help to make decisions.
+* Add necessary tests to make sure the refactoring is correct.
+* Non-bug refactoring is preferred to be done at the beginning of a milestone, it would be easier to find problems before the release.
+
+## Reviewing & Merging Suggestions
+
+* A refactoring PR shouldn't be kept open for a long time (usually 7 days), it should be reviewed as soon as possible.
+* A refactoring PR should be merged as soon as possible, it should not be blocked by other PRs.
+* If there is no objection from TOC, a refactoring PR could be merged after 7 days with one core member's approval (not the author).
```

Comment:
It's not just a suggestion, it also introduces a new rule, as a patch of https://github.com/go-gitea/gitea/blob/49919c636e4788a14f3762a0d9137ee2e0092790/CONTRIBUTING.md?plain=1#L122-L125

So we have to update `CONTRIBUTING.md`.

I would agree with:

1. If a PR has been ignored (no comment, no review) for a long time (more than 7 days), and the author or a maintainer believe it won't stand a long wait (like a refactoring PR), he/she can send a last call to TOC or all maintainers.
2. After another 7 days, if there is still zero approval, that means a polite refusal, then the PR will be closed to avoid wasting more time, so "a last call" has a cost, use it carefully.
3. However, if there is no objection from TOC, the PR could be merged with one core member's approval (not the author).

Reply:
TOC decides. Feel free to edit this PR directly.

Reply:
I think having a call out to TOC would be very useful - I think we've just started doing this in any case.